lib/vanagon/component/dsl.rb

```diff
@@ -416,6 +416,16 @@ def environment(*env) # rubocop:disable Metrics/AbcSize
         eos
       end
 
+      # Checks that the pkg_state is valid, returns vanagon error if invalid
+      #
+      # @param pkg_state [Array] array of pkg_state input to test
+      # @param function [String] the function that called
+      def check_pkg_state(pkg_state, method_name)
+        if pkg_state.empty? || !(pkg_state - ["install", "upgrade"]).empty?
```

Comment:
You could probably simplify this to a very, very simple `pkg_state.sort == %w[install upgrade]` and skip the `#empty?` and the `||` -- pkg_state **has** to be defined, and it should only ever be an Array; so sorting it and comparing it to the only allowed value should simplify this.

Reply:
Note that this will return `nil` in both the original implementation and my suggested revision -- this means it can't be used for direct comparisons or for any truthy operations, nor can it be included in any method chains. Consider this slight revision of my revision, which instead returns the original object using a [guard clause](http://www.rubydoc.info/github/bbatsov/rubocop/Rubocop/Cop/Style/GuardClause), or raises an error:

```ruby
def check_pkg_state(pkg_state, method_name)
  allowed_values = %w[install upgrade]
  return pkg_state unless pkg_state.sort != allowed_values.sort
  raise Vanagon::Error,
        "#{pkg_state} should be an array containing one or more of #{allowed_values} in ##{method_name}"
end
```

Reply:
So, reading the doc for this method, it looks like the argument`pkg_state` *must* be an Array, but that it can contain one or both of the allowed values? I didn't catch that before any of this but I think I did notice it the other night. I'd mentioned that `!(pkg_state - %w[install upgrade]).empty?` can be rewritten as `(pkg_state - %w[install upgrade]).any?`. Taking all of the above into account, that same guidance would make my revised revision look like this:

```ruby
def check_pkg_state(pkg_state, method_name)
  allowed_values = %w[install upgrade]
  return pkg_state if (pkg_state - allowed_values.sort).empty?
  raise Vanagon::Error,
        "#{pkg_state} should be an array containing one or more of #{allowed_values} in ##{method_name}"
end
```

Reply:
Tangent: that double `##` in the error message in my suggested revs. is intentional -- it'll print out the method name using the standard Ruby convention for an instance method:

```ruby
>> puts "#{pkg_state} should be an array containing one or more of #{allowed_values} in ##{method_name}"
["upgrade", "install"] should be an array containing one or more of ["install", "upgrade"] in #m_e_t_h_o_d_man
=> nil
```